### DISCUSSION TITLE
Allow having "anonymous" columns representing intermediate calculations but not surfaced in the final result

### DISCUSSION THREAD

**Discussion by norcalli** [duckdb/duckdb#14516] [Ideas]
Often times, for cleanliness, I just want to have some intermediate calculations where I give a name to them, but I don't want them to surface in the actual query result. To that end, I have to do a separate step where I strip out those columns but perhaps there could be a syntax for specifying anonymous columns.

e.g.

```sql
SELECT
  *,
   sum(shares) over () as anon.total_shares,
   total_shares * factor + case when shares / total_shares > 0.03 then 0 else 1 end as actual_result
```

we can express this as is by using `total_shares` as a column name, but the difference here is that I would only want to see `actual_result` added but not `total_shares`

**Comment by Tishj** [comment]
that's possible with subqueries, but I agree (and have raised this idea internally before) that it would be nice to provide a syntax way to make this easier

**Comment by Tishj** [comment]
The example I used in my idea:

current way is this:
```sql
query II
select * EXCLUDE a FROM (
    select <expr> a, function_one(a) c1, function_two(a) c2
);
----
c1    c2
```

my proposed idea, to add a TEMP keyword after a column to exclude it from the result
```sql
query II
select <expr> a TEMP, function_one(a) c1, function_two(a) c2;
----
c1    c2
```

**Reply by norcalli** [reply]
Nice, I think the `TEMP` syntax makes sense. I think that the ability to reference previous values is one of the more valuable syntax extensions that duckdb adds but without some way of excluding intermediate variables, it's not at its full potential/utility.

**Reply by norcalli** [reply]
As a side note, is there a way of qualifying whether you are referring to a newly defined column or one from a "from"?
I know for the from case, you can prepend with the table/subquery name, and I guess for the non-"from" case, you would just refer to it without qualification...?

I think this is part of why I used a syntax like `anon.` so that it's unambiguous what you're referring to when you refer to it, i.e. an intermediate variable.

**Reply by Tishj** [reply]
I don't think that's relevant here, that all works as is.
If the column is created in the current binding (as in, the select list) then it takes precedence over that same column in other bindings.

**Reply by soerenwolfers** [reply]
@norcalli there is a separate discussion for that at https://github.com/duckdb/duckdb/discussions/15658

**Comment by carlopi** [comment]
Another possible expansion could be having:
```sql
select <expr> a, function_one(a) c1, function_two(a) c2 EXCLUDE a;
```
parse / work correctly.

That could be equivalent to the already supported:
```sql
select * EXCLUDE a FROM (
    select <expr> a, function_one(a) c1, function_two(a) c2
);
```
but more terse and reusing existing building blocks.

**Reply by carlopi** [reply]
It's similar/derivative to @Tishj's `TEMP` proposal, possibly both can be considered.

**Reply by norcalli** [reply]
for multiple variables would it be this?

```sql
select <expr> a, <expr> b, function_one(a) c1, function_two(a) c2 EXCLUDE a, b;
```

**Reply by Tishj** [reply]
> but more terse and reusing existing building blocks.

I don't think that's true though, `EXCLUDE` is part of a star expression, meaning that this is valid:
```
D create table tbl(a varchar, b varchar, c varchar);
D select * EXCLUDE(b), b from tbl;
┌─────────┬─────────┬─────────┐
│    a    │    c    │    b    │
│ varchar │ varchar │ varchar │
├─────────┴─────────┴─────────┤
│           0 rows            │
└─────────────────────────────┘
```

Which would kind of break the expectance here, which is that EXCLUDE acts as a modifier on the whole selection list

**Reply by Tishj** [reply]
My proposal adds `TEMP` as a keyword to the expression, indicating that that column doesn't get propagated beyond the current binding

**Reply by norcalli** [reply]
Now that I think of it again, I think I agree with Tishj that TEMP on the column declaration site is less ambiguous syntactically
